bin/copy-production-modules.js

```diff
@@ -0,0 +1,124 @@
+const fs = require( 'fs' ); // eslint-disable-line  import/no-nodejs-modules
+const rcopy = require( 'recursive-copy' );
+const mkdirp = require( 'mkdirp' );
+const path = require( 'path' );
+const yargs = require( 'yargs' );
+
+const args = yargs
+	.usage( 'Usage: $0' )
+	.default( 'list', 'build/modules.json' )
+	.default( 'dest', 'build' )
+	.boolean( 'debug' ).argv;
+
+function debug( message ) {
+	args.debug && console.log( message );
+}
+
+try {
+	debug( 'reading modules from ' + args.list );
+	const externalModules = JSON.parse( fs.readFileSync( args.list, 'utf8' ) );
+	debug( 'bundle directly requests ' + externalModules.length + ' packages' );
+	const shippingPkgList = processPackages( externalModules );
+	shipDependencies( shippingPkgList );
+} catch ( err ) {
+	console.error( err );
+	process.exit( 1 );
+}
+
+function processPackages( pkgList ) {
+	const context = {
+		pkgList: [],
+		visitedFolders: new Set(),
+		folderStack: [ '.' ],
+		requiredBy: 'the bundle',
+		depth: 0,
+	};
+
+	for ( const pkgName of pkgList ) {
+		processPackage( pkgName, context );
+	}
+
+	return context.pkgList;
+}
+
+function processPackage( pkgName, context ) {
+	const { pkgList, folderStack, visitedFolders, requiredBy, depth } = context;
+
+	const pkgFolder = folderStack[ 0 ] + '/node_modules/' + pkgName;
+
+	// skip if the folder was already visited
+	if ( visitedFolders.has( pkgFolder ) ) {
+		return;
+	}
+
+	// mark the folder as visited
+	visitedFolders.add( pkgFolder );
+
+	// Verify that the package resolves to a directory in `node_modules/`, and that
+	// it is not a symlink to a monorepo packages. Such packages couldn't be shipped and
+	// must be bundled.
+	const pkgRealpath = path.relative( '.', fs.realpathSync( pkgFolder ) );
+	if ( ! pkgRealpath.startsWith( 'node_modules/' ) ) {
```

Comment:
What if server imports something that is in `client/node_modules`? Are those bundled?

Reply:
The server bundle is in `build/server.js`, both before and after this patch, so it doesn't see modules in `client/`. Its resolution paths are `build/node_modules`, `node_modules` and parents.

Modules in `client/` are visible only to webpack at build-time, when the resolution context are the source directories. Once a module is externalized, i.e., translated to a raw `require( 'foo' )` call executed by Node.js at runtime, this context is lost.

This could be a problem, because there's no guarantee that all externalized modules are always hoisted to the top `node_modules` directory. But this particular PR doesn't make it better or worse.

Reply:
Agreed this is an existing (and pretty bad) problem.

Reply:
There's one more little detail involved here: the `webpack-node-externals` plugin builds the list of externals by listing the root `node_modules` directory. If the sources request a `foo` package and the package is in `client/node_modules/foo` and not in `node_modules/foo`, the plugin won't identify `foo` as external and will let webpack bundle it.